### PR TITLE
Santosh Kumar Gajawada: Add source integration tests for the cdc and …

### PR DESCRIPTION
**Pull request by WeirdMagician**
…snapshot iterators

### Description
Source integration tests for the snapshot iterator and also the CDC iterator

Fixes # (issue)

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-s3/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:51]: Initially did the same but because of the linting issue reverted it back. Let me use the anonymous defer function to handle the source.Teardown() error for avoiding the linting issues.

Comment (WeirdMagician) [reply source/source_test.go:51]: Changes are done :)
Approve it if there are no improvements so i would merge it.

Later i will create a another issue to update the go test GitHub action to run the integration tests with the help of the configured GitHub secured variables.

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:99]: Here, we are creating a position ourselves, which is not bad. However, to have a better integration test, we should use the position returned by the source. So, I'd suggest following:
1. put some objects into the bucket
2. start the source
3. expect the objects to be read
4. stop and tear down the source
5. put some more objects into the bucket
6. start the source
7. expect the new objects to be read

File: source/source_test.go
Comment (hariso) [reply source/source_test.go:51]: Unless it's much work to add the GH secrets, it would be good if we add them before merging this PR. Seeing it green will make us all feel good about it.: )

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:99]: In the above steps tear down happened in the snapshot mode so starting again will re-trigger the snapshot again resulting duplicates. Below steps would fit the scenario.
1.put some objects into the bucket
2.start the source
3.expect the objects to be read
4.add an object again to change it to CDC mode 
5.stop and tear down the source
6.put some more objects into the bucket
7.start the source **with the last position previously stored.**
8.expect the new objects to be read

Comment (hariso) [reply source/source_test.go:99]: @WeirdMagician Exactly, I missed adding it to the list, even though I mentioned that we should use the position returned by the source.

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:51]: Completed Waiting for Approval to merge

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:99]: Completed Waiting for Approval to merge

**Comment by WeirdMagician** [comment]
Hi @hariso  GH secrets were added and the integration tests are not skipped and its GREEN now. Verify and share your observations.

File: source/source_test.go
Comment (hariso) [reply source/source_test.go:99]: At least in this test function, I don't think I see the steps we discussed above. I still see a single Open() call and a hard-coded position.

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:241]: What's the precision of maxLastModifiedDate? If it's milliseconds, then we should wait for only a millisecond.

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:307]: How about the payload?

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:292]: Same question as here: https://github.com/WeirdMagician/conduit-connector-google-cloudstorage/pull/10/files#r876060808

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:368]: This is actually the test for restarting a connector?

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:482]: This appears to be a magic value.

File: source/source_test.go
Comment (hariso) [inline source/source_test.go:455]: The test name feels a bit generic... What is the test about? Detecting deleted objects?

File: source/source.go
Comment (hariso) [inline source/source.go:120]: Why do we need to set it to nil?

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:495]: Here the Error is ErrBackoffRetry because there will be no records in CDC until the pollingPeriod. Here the default polling period is 1s and we didnt used any time.Sleep to skip this so an ErrBackoffRetry will be expected

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:701]: Sure

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:99]: For the steps we have discussed i have created a another testcase now it is rename to TestSource_CDCRestart

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:241]: Yes, It is updated now

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:256]: Seems the key is enough because we required whether it captured the exact action/event. Yes now i have also validating the content to make sure the data is updated correctly and same is captured.

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:307]: Seems the key is enough because we required whether it captured the exact action/event. Yes now i have also validating the content to make sure the data is updated correctly and same is captured.

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:292]: Updated

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:482]: It was a Past time, now i have changed testcase start time.

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go:455]: passed position type is CDC for the source_open method, with an insert and delete actions

File: source/source.go
Comment (WeirdMagician) [reply source/source.go:120]: Intention behind this is to make combined iterator invalidate and should not be used further after the teardown.
Because if the read is called after teardown then it should return an error of "combined iterator is not initialized".

Change the source read also to support when combined iterator is nil.

**Review by maha-hajja — COMMENTED** [review]
I noticed that a lot of lines logged in the test are from this line 
`fmt.Fprint(os.Stdout, "startCDC: iterator.Done\n")`
cdc_iterator.go line187 .. please change that into a trace log

File: source/iterator/cdc_iterator.go
Comment (hariso) [inline source/iterator/cdc_iterator.go:96]: It maybe better to log that we are attempting to stop the iterator, and then log that it was stopped after w.tomb.Dead().

File: source/source.go
Comment (hariso) [reply source/source.go:126]: The SDK guarantees that Teardown() will be the last call to the plugin, so from that point of view, setting it to nil is not really needed. If a function is called after Teardown() then something is messed up in the SDK or Conduit.

File: source/source_integration_test.go
Comment (hariso) [inline source/source_integration_test.go:112]: If it's not nil, why should it start from beginning?

File: source/source.go
Comment (WeirdMagician) [reply source/source.go:126]: So making combinedIterator nil would be a proactive step even if something messed up with the SDK. So should i continue with this ??

File: source/source_integration_test.go
Comment (WeirdMagician) [reply source/source_integration_test.go:112]: Sure i will elaborate the comment

File: source/source_integration_test.go
Comment (WeirdMagician) [reply source/source_integration_test.go:246]: ok

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: >  Here in this testcase the hard coded position is required to test the behavior of snapshot tests.

@WeirdMagician What is that test case actually? A restarted snapshot?

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: For some reason, I can't resolve this conversation. :+1: from me.

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: :+1:

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: :+1:

File: source/iterator/cdc_iterator.go
Comment (hariso) [reply source/iterator/cdc_iterator.go:98]: :+1:

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: :+1:

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: :+1:

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: :+1:

**Review by hariso — COMMENTED** [review]
@WeirdMagician I left a couple of minor comments, I believe we can merge this very soon.

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go]: May i know is anything left here ??
Because currently iam using the defer block for source teardown and even add the GH secrets to run the integration tests via github actions

Comment (WeirdMagician) [reply source/source_test.go]: May i know is anything left here ??
Because currently we are using the defer block for source tear down and even add the GitHub secrets to run the integration tests via GitHub actions

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go]: Yes this behavior will happen when it got restarted from a non-nil snapshot position.
Here in this particular testcase we are not testing the entire connector restart use-case. Instead we are testing the behavior post restart called (SnapshotStartFromNonNilPosition).

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go]: Ok Now we are validating the payload too.

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go]: Got it. I will make sure that it is ErrBackoffRetry, if not the test will fail

File: source/source.go
Comment (WeirdMagician) [reply source/source.go:126]: Got it. Let me remove it

**Comment by WeirdMagician** [comment]
@hariso i have fixed minor comments.

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: It would be good to have a full test case, i.e. one where you start a snapshot, stop it, then start again. But we can do that in a separate PR (if you agree with the idea in the first place).

File: source/source_test.go
Comment (hariso) [reply source/source_test.go]: I think we are good here. :+1:

**Review by hariso — APPROVED** [review]

File: source/source_test.go
Comment (WeirdMagician) [reply source/source_test.go]: Add in the existing issue #23

**Pull request merged**
